components/script/dom/vrdisplay.rs

```diff
@@ -0,0 +1,618 @@
+                if let Ok(depth) = raf_receiver.recv().unwrap() {
+                    near = depth.0;
+                    far = depth.1;
+                } else {
+                    // Stop thread
+                    // ExitPresent called or some error happened
+                    return;
+                }
+            }
+        }).expect("Thread spawning failed");
+    }
+
+    fn stop_present(&self) {
+        self.presenting.set(false);
+        *self.frame_data_receiver.borrow_mut() = None;
+
+        let api_sender = self.layer_ctx.get().unwrap().ipc_renderer();
+        let display_id = self.display.borrow().0.display_id;
+        let msg = VRCompositorCommand::Release(display_id);
+        api_sender.send(CanvasMsg::WebVR(msg)).unwrap();
+    }
+
+    // Only called when the JSContext is destroyed while presenting
+    // In this case we don't want to wait for WebVR Thread response
+    fn force_stop_present(&self) {
+        if let Some(wevbr_sender) = self.webvr_thread() {
+            wevbr_sender.send(WebVRMsg::ExitPresent(self.global().pipeline_id(),
+                                                    self.display.borrow().0.display_id,
+                                                    None))
+                                                    .unwrap();
+        }
+        self.stop_present();
+    }
+
+    fn sync_frame_data(&self) {
+        let status = if let Some(receiver) = self.frame_data_receiver.borrow().as_ref() {
+            match receiver.recv().unwrap() {
+                Ok(bytes) => {
+                    self.frame_data.borrow_mut().0 = webvr::VRFrameData::from_bytes(&bytes[..]);
+                    VRFrameDataStatus::Synced
+                },
+                Err(()) => {
+                    VRFrameDataStatus::Exit
+                }
+            }
+        } else {
+            VRFrameDataStatus::Exit
+        };
+
+        self.frame_data_status.set(status);
+    }
+
+    fn handle_raf(&self, end_sender: &mpsc::Sender<Result<(f64, f64), ()>>) {
+        self.frame_data_status.set(VRFrameDataStatus::Waiting);
+
+        let mut callbacks = mem::replace(&mut *self.raf_callback_list.borrow_mut(), vec![]);
+        let now = self.global().as_window().Performance().Now();
+
+        for (_, callback) in callbacks.drain(..) {
+            if let Some(callback) = callback {
+                let _ = callback.Call__(Finite::wrap(*now), ExceptionHandling::Report);
+            }
+        }
+
+        if self.frame_data_status.get() == VRFrameDataStatus::Waiting {
+            // User didn't call getFrameData while presenting
+            // Show a warning as the WebVR Spec recommends
+            warn!("WebVR: You should call GetFrameData while presenting");
+            self.sync_frame_data();
+        }
+
+        match self.frame_data_status.get() {
+            VRFrameDataStatus::Synced => {
+                end_sender.send(Ok((self.depth_near.get(), self.depth_far.get()))).unwrap();
+            },
```

Comment:
Can you use explicit matching here? It's easier to see how this finishes the RAF thread.